### PR TITLE
Create API to get the state of the message in a backend. Fix #107

### LINKED ISSUE #107: Add a remoulade API
We should be able to get data for the dashboard

Use flask https://palletsprojects.com/p/flask/ as the web framework.

No authentication is needed as it will be inside the kubernetes cluster

For now, it will only access the data of the status middleware #106 

## Checklist

* [ ] Create a method in `MessageState` to get all the states (using [scan](https://redis.io/commands/scan))
* [ ] Create a flask Http server
* [ ] Create a route `/messages/states` that returns all the states stored in `MessageState` middleware, Also add a get parameter `name` to filter on the state name.

### PR DESCRIPTION
**Pull request by xdanielsb**
- code tests
- apply checkers

### CODE REVIEW

File: remoulade/api/main.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+      Return  the message  with a certain id
+    """
+
+    error = {"IdNotFound": {"errorCode": "idNotFound", "errorMessage": "Id not found"}}
+
+    def get(self, message_id):
+        data = jsonify(admin.get(message_id))
+        if data is None:
+            return make_response(jsonify(self.error["IdNotFound"]), 400)
+        return make_response(data, 200)
```
[the remark this answers is not in the document]
Comment (antoinerabany) [reply remoulade/api/main.py:35]: https://flask.palletsprojects.com/en/1.1.x/quickstart/#apis-with-json

File: remoulade/api/app/__init__.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+""" Creating the instance of the app """
+from flask import Flask
+
+from remoulade.state import State
+from typing import Dict, List
+from .config import Config
+
+
+class MessageAdmin:
```
Comment (antoinerabany) [inline remoulade/api/app/__init__.py:9]: I would prefer you do not use this object, I make things more complex than it should be. We should read the code directly in the route, making things more clear.

File: remoulade/api/app/__init__.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    def __init__(self):
+        self.app = Flask(__name__)
+        self.app.config.from_object(Config)
+
+    def get_app(self):
+        return self.app
+
+    def get(self, message_id) -> Dict[str, str]:
+        res = self.app.config["backend"].get_state(message_id)
```
Comment (antoinerabany) [inline remoulade/api/app/__init__.py:18]: I would prefer that you get the state backend via the remouade broker.
check `get_cancel_backend` and do something similar for state.

File: remoulade/api/main.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+
+class MessageStateAPI(MethodView):
```
Comment (antoinerabany) [inline remoulade/api/main.py:38]: I would prefer you use decorator like app.route(), it's more simple.

File: remoulade/api/main.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    """
+      /messages/state/<state>
```
Comment (antoinerabany) [inline remoulade/api/main.py:40]: there should be only two routes  `/messages/states` and `/messages/state/{message_id}`

You should pass state as GET parameter. (/messages/states?state=abc)

```
from flask import request
state = request.args.get('state')
```

File: remoulade/api/app/config.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
+""" Configuration of the app """
+import os
+
+
+class Config:
```
Comment (antoinerabany) [inline remoulade/api/app/config.py:5]: do you really need this config ?

File: remoulade/api/main.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+""" This file describe the API to get the state of messages """
+from flask import jsonify, make_response
+from flask.views import MethodView
+
+from remoulade.state import StateNamesEnum
+
+from .app import MessageAdmin
+
+admin = MessageAdmin()
+app = admin.get_app()
```
Comment (antoinerabany) [inline remoulade/api/main.py:10]: `app = Flask(__name__)` is enough right ?

File: tests/state/test_state_api.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+from remoulade.state import StateNamesEnum
+
+
+@pytest.fixture
+def api_client(state_middleware):
+    app.config["TESTING"] = True
+    app.config["backend"] = state_middleware.backend
+    with app.test_client() as client:
+        yield client
+    app.config["TESTING"] = False
```
Comment (antoinerabany) [inline tests/state/test_state_api.py:13]: I think we don't really need `app.config["TESTING"] = False`. As the real falsk app will never be started after the end of a test.

File: tests/state/test_state_api.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                assert state_enum.name.encode() not in res.data
+
+    def test_invalid_state_message(self, stub_broker, do_work, api_client):
+        do_work.send()
+        res = api_client.get("/messages/states/?name=invalid_state")
+        assert b"Bad Request" in res.data
+
+    def test_success_message(self, stub_broker, stub_worker, state_middleware, do_work, api_client):
+        do_work.send()
+        stub_broker.join(do_work.queue_name)
```
Comment (antoinerabany) [inline tests/state/test_state_api.py:41]: Do not create a state by enqueing a message it has already been tested elsewhere, you can add it directly with state_middleware.backend like this.

```
                for message_id in range(1500):
            state_middleware.backend.set_state(State(message_id, StateNamesEnum.Success, [], {}), 1000)

        res = api_client.get("/messages/states/")
        assert len(res.json) == 1500
```

I wanted to test we were returning more than 1000 states.

File: tox.ini [hunk 10]
```diff
@@ -8,6 +8,8 @@ passenv=
   CI
 
 [testenv:test]
+deps=
+  flask>=1.1,<2
```
Comment (antoinerabany) [inline tox.ini:12]: If you take a look at the circle ci file, you see it does `pip install '.[dev]' tox` to intsall the dependencies.

So as flask is already inclueded in dev dependencies (look at setup.py line 35), it should not be needed.

Comment (antoinerabany) [reply tox.ini:12]: could you send me the Ci where it fails when remove it ?

File: remoulade/api/main.py [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+from flask import Flask, jsonify, make_response, request
+from werkzeug.exceptions import BadRequest, NotFound
+
+import remoulade
+from remoulade.state import StateNamesEnum
+
+app = Flask(__name__)
+
+
+@app.route("/messages/states/")
```
Comment (antoinerabany) [inline remoulade/api/main.py:11]: generally there is no / at the end of routes, so `/messages/states` would be better

File: remoulade/api/main.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+        else:
+            if not any(state_enum for state_enum in StateNamesEnum if state_enum.name.lower() == name_state.lower()):
+                raise BadRequest("{} state is not defined.".format(name_state))
+            data = [state.asdict() for state in backend.get_n_states() if state.name.lower() == name_state.lower()]
+        return make_response(jsonify(data), 200)
+    except BadRequest as e:
+        return e
+
+
+@app.route("/mesassegs/state/<message_id>")
```
Comment (antoinerabany) [inline remoulade/api/main.py:28]: there is a typo in the message name

File: remoulade/api/main.py [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+    try:
+        data = None
+        if name_state is None:
+            data = [d for d in backend.get_n_states()]
+        else:
+            if not any(state_enum for state_enum in StateNamesEnum if state_enum.name.lower() == name_state.lower()):
+                raise BadRequest("{} state is not defined.".format(name_state))
+            data = [state.asdict() for state in backend.get_n_states() if state.name.lower() == name_state.lower()]
+        return make_response(jsonify(data), 200)
+    except BadRequest as e:
```
Comment (antoinerabany) [inline remoulade/api/main.py:24]: do not catch the error here but in a error handler like this 
```
@app.errorhandler(HTTPException)
def http_exception(e):
    return {"error": str(e)}, e.code
```

as HTTPException is a superclass of BadRequest and NotFound, we will send a nice json when we get these exceptions

File: remoulade/api/main.py [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+
+@app.route("/mesassegs/state/<message_id>")
+def get_state(message_id):
+    try:
+        backend = remoulade.get_broker().get_state_backend()
+        data = backend.get_state(message_id)
+        if data is None:
+            raise NotFound("message_id = {} does not exist".format(message_id))
+        return make_response(data, 200)
```
Comment (antoinerabany) [inline remoulade/api/main.py:35]: return data.asdict() and add a test for this route, it was broken

File: remoulade/api/main.py [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+
+import remoulade
+from remoulade.state import StateNamesEnum
+
+app = Flask(__name__)
+
+
+@app.route("/messages/states/")
+def get_states():
+    name_state = request.args.get("name")
```
Comment (antoinerabany) [inline remoulade/api/main.py:13]: I tried to make a simpler implementation of this method: 
```
name_state = request.args.get("name")
    backend = remoulade.get_broker().get_state_backend()
    data = [s.asdict() for s in backend.get_states()]
    if name_state is not None:
        if name_state.lower() not in [state_enum.name.lower() for state_enum in StateNamesEnum]:
            raise BadRequest("{} state is not defined.".format(name_state))
        data = [s for s in data if s["name"] == name_state.lower()]
    return {'result': data}
```
I try to call only once `get_states` to keep things clear.
And it's more pythonic to `if a not in [1, 2]` as it's exaclty what you want to test, if a value is in a array

File: tests/state/test_state_api.py (hunk 9 above)
Comment (antoinerabany) [reply tests/state/test_state_api.py:41]: with this methods you could add some more tests

File: tests/state/test_state_api.py [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+        assert res.data == b"[]\n"
+
+    def test_invalid_url(self, stub_broker, state_middleware, api_client):
+        res = api_client.get("/invalid_url/")
+        assert res.status_code == 404
+
+    def test_pending_message(self, stub_broker, state_middleware, do_work, api_client):
+        do_work.send()
+        res = api_client.get("/messages/states/")
+        for state_enum in StateNamesEnum:
```
Comment (antoinerabany) [inline tests/state/test_state_api.py:28]: you should not have to do these for loop to get an enum. It's because of (in states.enum)
```
    def __get__(self, instance, owner):
        return self.value
```
We should not overide how enum works.

Sorry, I thougth it was good but after som though it's not.

I made some changes to the enums:
```
class StateNamesEnum(Enum):
    """
    Contains the possible states that can have a message:
    """

    Started = "Started"  # a `Message` that has not been processed
    Pending = "Pending"  # a `Message` that has been enqueued
    Skipped = "Skipped"  # a `Message` that has been skipped
    Canceled = "Canceled"  # a `Message` that has been canceled
    Failure = "Failure"  # a `Message` that has been processed and raised an Exception
    Success = "Success"  # a `Message` that has been processed and does not raise an Exception


#: A type alias representing states in the database
class State(namedtuple("State", ("message_id", "name", "args", "kwargs"))):
    """Catalog Class, it storages the state
        Parameters:
            name: Name of the state
            args: List of arguments in the state(name)
    """

    def __new__(cls, message_id, name, args, kwargs):
        if name not in [state for state in StateNamesEnum]:
            raise InvalidStateError("The {} State is not defined".format(name))
        return super().__new__(cls, message_id, name, args, kwargs)

    @classmethod
    def from_dict(cls, dict):
        dict["name"] = StateNamesEnum(dict["name"])
        return cls(**dict)

    def asdict(self):
        return {**self._asdict(), "name": self.name.value}
```

This way you will be able to do:
`assert StateNamesEnum.Pending.name.encode() in res.data`

File: tests/state/test_state_api.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+        assert res.status_code == 200
+        assert len(res.json["result"]) == 0
+
+    def test_invalid_url(self, stub_broker, state_middleware, api_client):
+        res = api_client.get("/invalid_url/")
+        assert res.status_code == 404
+
+    def test_pending_message(self, stub_broker, state_middleware, api_client):
+        state_middleware.backend.set_state(State("id", StateNamesEnum.Pending, [], {}), ttl=1000)
+        res = api_client.get("/messages/states")
```
Comment (antoinerabany) [inline tests/state/test_state_api.py:27]: Can you add like 3 or 4 states with different status and test for the full json ?

File: remoulade/common.py [hunk 18]
```diff
@@ -147,3 +149,11 @@ def flatten(l):
             yield from flatten(el)
         else:
             yield el
+
+
+def chunk(iterable: Iterable, size: int) -> Iterator[List]:
```
Comment (antoinerabany) [inline remoulade/common.py:154]: can you add a small docstring for this fucntion ?

File: remoulade/api/main.py [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    return {"result": data}
+
+
+@app.route("/messages/state/<message_id>")
+def get_state(message_id):
+    backend = remoulade.get_broker().get_state_backend()
+    data = backend.get_state(message_id)
+    if data is None:
+        raise NotFound("message_id = {} does not exist".format(message_id))
+    return {"result": data.asdict()}, 200
```
Comment (antoinerabany) [inline remoulade/api/main.py:30]: You don't need result here as the returned data is not a list.

File: remoulade/api/main.py [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+
+
+@app.route("/messages/states")
+def get_states():
+    name_state = request.args.get("name")
+    backend = remoulade.get_broker().get_state_backend()
+    data = [s.asdict() for s in backend.get_states()]
+    if name_state is not None:
+        name_state = name_state.lower()
+        if not any(state_enum for state_enum in StateNamesEnum if state_enum.name.lower() == name_state.lower()):
```
Comment (antoinerabany) [inline remoulade/api/main.py:18]: use the if name_state not in ... instead of if any.

File: remoulade/state/backend.py [hunk 21]
```diff
@@ -17,29 +18,27 @@ class StateNamesEnum(Enum):
...
     """Catalog Class, it storages the state
         Parameters:
             name: Name of the state
             args: List of arguments in the state(name)
     """
 
-    def __new__(cls, name, args, kwargs):
-        if not any(state for state in StateNamesEnum if state.name == name):
+    def __new__(cls, message_id, name, args, kwargs):
+        if name not in list(StateNamesEnum):
```
Comment (antoinerabany) [inline remoulade/state/backend.py:31]: Why did you add list ?

File: remoulade/api/main.py [hunk 22]
```diff
@@ -0,0 +1,35 @@
+""" This file describe the API to get the state of messages """
+from flask import Flask, request
+from werkzeug.exceptions import BadRequest, HTTPException, NotFound
```
Comment (antoinerabany) [inline remoulade/api/main.py:3]: could you add `"remoulade.api"` to the package line in setup.py line 65 ? It make it possible to import the given sup package.

And add app to remoulade/api/__init__.py, don't forget to add it to __all__ (look at other inits).

Then could you write le little changelog ? 

Thanks !

File: remoulade/state/backend.py (hunk 21 above)
Comment (xdanielsb) [reply remoulade/state/backend.py:31]: there was a warning, using non-Enums in containment checks will raise TypeError in Python 3.8

File: docs/source/changelog.rst [hunk 23]
```diff
@@ -8,6 +8,22 @@ All notable changes to this project will be documented in this file.
 `Unreleased`_
 -------------
 
+`0.21.0`_ -- (unreleased)
+-------------------------
+
+Added
+^^^^^
+* Error |NoStateBackend| raised when is tried to access a |StateBackend| in a broker without it
+* Method ``chunk`` in ``commons.py`` return an iterable over a list of length ``size``
```
Comment (antoinerabany) [inline docs/source/changelog.rst:17]: As you can see in setup.py, remoulade.common is not included in the package. that mean all functions inside it are private to remoulade. So no need to add it to the changelog.

File: remoulade/api/main.py [hunk 24]
```diff
@@ -0,0 +1,37 @@
...
+
+@app.route("/messages/states")
+def get_states():
+    name_state = request.args.get("name")
+    backend = remoulade.get_broker().get_state_backend()
+    data = [s.asdict() for s in backend.get_states()]
+    if name_state is not None:
+        name_state = name_state.lower().capitalize()
+        try:
+            state = StateNamesEnum(name_state)
```
Comment (antoinerabany) [inline remoulade/api/main.py:19]: Nice ! that's very pythonic !

File: remoulade/api/main.py [hunk 25]
```diff
@@ -0,0 +1,37 @@
...
+    return {"result": data}
+
+
+@app.route("/messages/state/<message_id>")
+def get_state(message_id):
+    backend = remoulade.get_broker().get_state_backend()
+    data = backend.get_state(message_id)
+    if data is None:
+        raise NotFound("message_id = {} does not exist".format(message_id))
+    return data.asdict(), 200
```
Comment (antoinerabany) [inline remoulade/api/main.py:32]: no need for the 200. as it's the default.

File: tests/state/test_state_api.py [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+                choice(list(StateNamesEnum)),  # random state
+                sample(range(1, 10), randint(0, 5)),  # random args
+                {str(i): str(i) for i in range(randint(0, 5))},  # random kwargs
+            )
+            random_states.append(random_state.asdict())
+            state_middleware.backend.set_state(random_state, ttl=1000)
+
+        # check storage
+        res = api_client.get("/messages/states")
+        states = res.json["result"]
```
Comment (antoinerabany) [inline tests/state/test_state_api.py:61]: the test are perfect now !
